projects/l2_pooling/noise_tolerance.py

```diff
@@ -0,0 +1,809 @@
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2016, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+from collections import defaultdict
+import math
+import random
+import os
+import time
+
+import matplotlib.pyplot as plt
+
+from htmresearch.algorithms.column_pooler import ColumnPooler
+from htmresearch.frameworks.layers.sensor_placement import greedySensorPositions
+
+L4_CELL_COUNT = 8*1024
+
+
+def createRandomObjectDescriptions(numObjects,
+                                   numLocationsPerObject,
+                                   featurePool=("A", "B", "C")):
+  """
+  Returns {"Object 1": [(0, "C"), (1, "B"), (2, "C"), ...],
+           "Object 2": [(0, "C"), (1, "A"), (2, "B"), ...]}
+  """
+  return dict(("Object %d" % i,
+               zip(xrange(numLocationsPerObject),
+                   [random.choice(featurePool)
+                    for _ in xrange(numLocationsPerObject)]))
+              for i in xrange(1, numObjects + 1))
+
+
+def noisy(pattern, noiseLevel, totalNumCells):
+  n = int(noiseLevel * 40)
+
+  noised = set(pattern)
+
+  noised.difference_update(random.sample(noised, n))
+
+  for _ in xrange(n):
+    while True:
+      v = random.randint(0, totalNumCells - 1)
+      if v not in pattern and v not in noised:
+        noised.add(v)
+        break
+
+  return noised
+
+
+def doExperiment(numColumns, l2Overrides, objectDescriptions, noiseFn,
+                 numInitialTraversals):
+
+  # For each column, keep a mapping from feature-location names to their SDRs
+  layer4sdr = lambda : set(random.sample(xrange(L4_CELL_COUNT), 40))
+  featureLocationSDRs = [defaultdict(layer4sdr) for _ in xrange(numColumns)]
+
+  poolers = [ColumnPooler(inputWidth=L4_CELL_COUNT,
+                          lateralInputWidths=[4096]*(numColumns-1),
+                          **l2Overrides)
+             for _ in xrange(numColumns)]
+
+  # Learn the objects
+  objectL2Representations = {}
+  for objectName, featureLocations in  objectDescriptions.iteritems():
+    for featureLocationName in featureLocations:
+      for _ in xrange(10):
```

Comment:
Hmm. 3 is risky, since it will result in distal permanences of exactly 0.6, so a single `synPermDistalDec` will put it below threshold. (since distal learning only occurs on the second and third time step). Yeah, 4 should be enough. I did more because this script isn't really testing the learning rules, so it avoids all risk of accidentally falling below the connected threshold, and focuses on inference. But I'm cool with doing 4 if we want to do the same thing across all our scripts. And, of course, it's faster.